index.src.html

```diff
@@ -2260,14 +2290,805 @@
 
         NOTE: User Agents are encouraged to avoid expensive copies of large
             objects (for instance, {{VideoFrame}} pixel data). Frame types are
-            immutable, so the above step may be implemented using memory sharing
-            techniques such as reference counting.
+            immutable, so the above step may be implemented using memory
+            sharing techniques such as reference counting.
 
     3. Return |cloneFrame|.
 
   </dd>
 </dl>
 
+Image Decoding {#image-decoding}
+====================================
+
+Background {#image-decoding-background}
+-------------------------------------
+
+This section is non-normative.
+
+Image codec definitions are typically accompanied by a definition for a
+corresponding file format. Hence image decoders often perform both duties of
+unpacking (demuxing) as well as decoding the encoded image data. The WebCodecs
+{{ImageDecoder}} follows this pattern, which motivates an interface design that
+is notably different from that of {{VideoDecoder}} and {{AudioDecoder}}.
+
+In spite of these differences, {{ImageDecoder}} uses the same
+[=codec processing model=] as the other codec interfaces. Additionally,
+{{ImageDecoder}} uses the {{VideoFrame}} interface to describe decoded outputs.
+
+ImageDecoder Interface {#imagedecoder-interface}
+------------------------------------------------
+
+<pre class='idl'>
+<xmp>
+[Exposed=(Window,DedicatedWorker)]
+interface ImageDecoder {
+  constructor(ImageDecoderInit init);
+
+  readonly attribute boolean complete;
```

Comment:
This needs an event associated with the value change. It was also mentioned a few minutes ago in the editor's call that this could be removed ?

Reply:
It could be, but it feels ugly if we do. Right now it's used to do this when ImageDecoder is used with a ReadableStream:

```
fetch('example_image').then(r => {
  let decoder =
      new ImageDecoder({data: r.body, type: r.headers.get("Content-Type")});

  let currentFrameIndex = 0;
  function decodeMoreFrames(result) {
    if (decoder.complete)
      return;
    decoder.decode({frameIndex: ++currentFrameIndex}).then(decodeMoreFrames);
  }

  decoder.decode({frameIndex: currentFrameIndex}).then(decodeMoreFrames);
});
```

The alternative is:
```
  function decodeMoreFrames(result) {
    decoder.decode({frameIndex: ++currentFrameIndex})
        .then(decodeMoreFrames)
        .catch(e => {
          if (e instanceof RangeError)
            return;
          throw e;
        });
  }
```

I'm okay with adding another promise-property or event, but we'd still need the boolean for synchronous tests. How does a promise-property like 'completed' sound? Too close to the boolean?

Reply:
I think the decode-until-exception is workable, but unusual. I think a promise is better than an event here, per https://www.w3.org/2001/tag/doc/promises-guide#one-time-events. 

`complete` and `completed` are pretty close, but I don't hate it. I was tempted to name the promise completePromise, but I'm not finding [precedent for that style](https://source.chromium.org/search?q=f:idl$%20%22Promise;%22%20%20-f:test%20f:blink&sq=). Another option would be to name the boolen isComplete. Lots of [precedent for that](https://source.chromium.org/search?q=f:idl$%20%22boolean%20is%22%20-f:ppapi%20f:blink&sq=) . @padenot, do you like that idea?

Reply:
I agree that using exceptions for control flow in perfectly normal (non-error) circumstances is not great, and is generally frowned upon in software engineering. A promise property sounds like a compelling option to me.

In general (and there are notable exceptions, with clear justifications), a property has to have something event-based to avoid any temptation to busy loop (even though it wouldn't do anything here).

I was thinking that maybe a generator / iterator pattern would lend itself well to this task, but the current style allows random access, and a generator / iterator wouldn't (afaik).

Reply:
FWIW, &lt;img&gt; only has a complete property: https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/complete --- but the example on MDN would clearly benefit from a completed promise :)

Reply:
I went with completed, but can change later if there are objections.

Reply:
Latest commit adds that promise. Leaving open for @padenot to resolve if this addresses his concern.

Reply:
Yep!